### PR TITLE
Add support to prefix for envFrom

### LINKED ISSUE #38604: Kubernetes extension: Support env vars from secrets with a prefix
### Description

Right now you can apply a Secret (or ConfigMap) as environment variables like this:

`quarkus.kubernetes.env.secrets=my-secret` 

This would map all keys of the Secret (or ConfigMap) to the environment of the Kubernetes deployment manifest:

```
envFrom:
  - secretRef:
       name: my-secret
```

However, especially when you're dealing with more than - perhaps - one datasource, it would be great to specify also a prefix. Kubernetes supports adding a prefix to imported keys from Secret / ConfigMap. 

```
envFrom:
  - secretRef:
    name: my-secret
  prefix: DB1_
```
If `my-secret` contains a key called `user` it would be mapped to `DB1_user`, which you then could use to specify your datasource properties in `application.properties`:

```
%prod.quarkus.datasource.username=${DB1_user:wanja}
```

It would be great if Quarkus could allow the specification of a prefix in`application.properties` for example by using this grammar:

`quarkus.kubernetes.env.secrets=my-secret[DB1_],my-other-secret[DB2_]`

Or by introducing another sub key of `env` called `prefix`:

`quarkus.kubernetes.env.prefix.my-secret=DB1_`

### Implementation ideas

_No response_

### PR DESCRIPTION
**Pull request by mcruzdev**
- Fixes #38604

We need to wait the release of new https://github.com/dekorateio/dekorate version

### CODE REVIEW

**Comment by gsmet** [comment]
Converting to draft as it's using a snapshot version.

**Comment by mcruzdev** [comment]
Hi @gsmet I think that we can review now, thank you @iocanel for release the dekorate version :)

**Comment by mcruzdev** [comment]
I will fix the import sort soon...

File: bom/application/pom.xml [hunk 1]
```diff
@@ -164,7 +164,7 @@
         <kotlin.coroutine.version>1.8.0</kotlin.coroutine.version>
         <azure.toolkit-lib.version>0.27.0</azure.toolkit-lib.version>
         <kotlin-serialization.version>1.6.2</kotlin-serialization.version>
-        <dekorate.version>4.1.2</dekorate.version> <!-- Please check with Java Operator SDK team before updating -->
+        <dekorate.version>4.1.3</dekorate.version> <!-- Please check with Java Operator SDK team before updating -->
```
Comment (mcruzdev) [inline bom/application/pom.xml:167]: @gsmet, should this change be on separated pull request?

**Comment by mcruzdev** [comment]
Hi @geoand, @iocanel, how are you? Could you review it, I think that is ok

**Review by iocanel — CHANGES_REQUESTED** [review]
I am not a big fan of the convention approach that is used to set the prefix.
Someone looking at this config will have a really hard time understanding what this is about.
Unfortunately, env var configuration is already pretty hard and confusing already, so we should try our best not to add to that.

So, I would like us to be explicit when setting the `prefix` (e.g. the literal `prefix` being part of the property name). 

Maybe something like:

```
quarkus.kubernetes.env.prefix."MY_PREFIX_".fromSecret=my-secret
quarkus.kubernetes.env.prefix."OTHER_PREFIX_".fromConfigMap=my-cm
```

This rhymes with how we do key mapping, so the pattern is already familiar to users. AFAIR @metacosm implemented that, so I would like his input as well.

**Comment by metacosm** [comment]
I think the best way to do this would be to use something like:
```
quarkus.kubernetes.env.secrets.prefix."my-secret"=DB1_
```
This would register the `DB1_` prefix with the `my-secret` secret and presumably use that prefix everywhere the secret is used. 
The same should probably be implemented for ConfigMaps.

**Comment by mcruzdev** [comment]
I think @metacosm suggestion more simple, wdyt @iocanel.

**Comment by mcruzdev** [comment]
> I think the best way to do this would be to use something like:
> 
> ```
> quarkus.kubernetes.env.secrets.prefix."my-secret"=DB1_
> ```
> 
> This would register the `DB1_` prefix with the `my-secret` secret and presumably use that prefix everywhere the secret is used. The same should probably be implemented for ConfigMaps.

Hi @metacosm, to do this one will be necessary to change the EnvVarsConfig POJO.

Removing the `List<String> secrets` and adding something like `mapping`  ` Map<String, EnvVarFromKeyConfig> mapping;
`.

**Comment by metacosm** [comment]
> Hi @metacosm, to do this one will be necessary to change the EnvVarsConfig POJO.
> 
> Removing the `List<String> secrets` and adding something like `mapping` `Map<String, EnvVarFromKeyConfig> mapping;`.
> 
> The secrets will need a change:
> 
> From: `quarkus.kubernetes.env.secrets=my-secret,my-other-secret` to:
> 
> ```ini
> quarkus.kubernetes.env.secrets.my-secret.name=my-secret
> quarkus.kubernetes.env.secrets.my-secret.prefix=DBZ_
> ```
> 
> or you can follow what @iocanel suggested!
> 
> Let me know if I am wrong 😃

You're right.

Maybe we could use the `mapping` configuration to our advantage here, though, with something like:

```properties
quarkus.kubernetes.env.mapping."DB1_".as-prefix-for-secret=my-secret
quarkus.kubernetes.env.mapping."CM1_".as-prefix-for-configmap=my-cm
```

**Comment by metacosm** [comment]
> 
> It means that to set the prefix we need to set the with-key. two properties to set the prefix, what do you think about @iocanel suggestion?

You're right again :)

> 
> ```ini
> quarkus.kubernetes.env.prefix."MY_PREFIX_".fromSecret=my-secret
> ```

I guess that's indeed the best solution but then I'd do: 

```properties
quarkus.kubernetes.env.prefixes."MY_PREFIX_".for-secret=my-secret
```

If we wanted to be more fluent, we could do:

```properties
quarkus.kubernetes.env.using."MY_PREFIX_".as-prefix-for-secret=my-secret
```

but I guess that would be less legible on the `ConfigGroup` in the code…

**Comment by geoand** [comment]
@iocanel can you have a look at this?

Thanks

**Review by iocanel — APPROVED** [review]
LGTM

File: bom/application/pom.xml (hunk 1 above)
Comment (geoand) [reply bom/application/pom.xml:157]: @metacosm I assume the version bump is fine with you?

Comment (metacosm) [reply bom/application/pom.xml:157]: I actually don't know and didn't get a chance to check… That version of dekorate uses an older version than the one used in the Kubernetes client and I have no idea about compatibility. I would have preferred this upgrade to be isolated from this PR if at all possible and be consulted before merging it, as is mentioned in the POM file.

Comment (geoand) [reply bom/application/pom.xml:157]: Fair enough. 

When you get please chance, please check the version and see if everything works as expected.

Thanks!

**Pull request merged**